tree/ntuple/v7/inc/ROOT/RField.hxx

```diff
@@ -174,6 +174,83 @@ public:
+      std::size_t fValueSize = 0;         ///< Cached copy of fField->GetValueSize()
+      std::size_t fCapacity = 0;          ///< The size of the array memory block in number of values
+      std::size_t fSize = 0;              ///< The number of available values in the array (provided their mask is set)
+      std::unique_ptr<bool[]> fMaskAvail; ///< Masks invalid values in the array
+      std::size_t fNValidValues = 0;      ///< The sum of non-zero elements in the fMask
+      RClusterIndex fFirstIndex;          ///< Index of the first value of the array
+
+      void ReleaseValues();
+      /// Sets a new range for the bulk. If there is enough capacity, the fValues array will be reused.
+      /// Otherwise a new array is allocated. After reset, fMaskAvail is false for all values.
+      void Reset(const RClusterIndex &firstIndex, std::size_t size);
+      void CountValidValues();
+
+      bool ContainsRange(const RClusterIndex &firstIndex, std::size_t size) const
+      {
+         if (firstIndex.GetClusterId() != fFirstIndex.GetClusterId())
+            return false;
+         return (firstIndex.GetIndex() >= fFirstIndex.GetIndex()) &&
+                ((firstIndex.GetIndex() + size) <= (fFirstIndex.GetIndex() + fSize));
+      }
+
+      void *GetValuePtrAt(std::size_t idx) const { return &fValues[idx * fValueSize]; }
+
+      explicit RBulk(RFieldBase *field) : fField(field), fValueSize(field->GetValueSize()) {}
+
+   public:
+      ~RBulk();
+      RBulk(const RBulk &) = delete;
+      RBulk(RBulk &&other);
+      RBulk &operator=(RBulk &&other);
+
+      /// Reads 'size' values from the associated field, starting from 'firstIndex'. Note that the index is given
+      /// relative to a certain cluster. The return value points to the array of read objects.
+      /// The 'maskReq' parameter is a bool array of at least 'size' elements. Only objects for which the mask is
+      /// true are guaranteed to be read in the returned value array.
+      void *ReadBulk(const RClusterIndex &firstIndex, const bool *maskReq, std::size_t size)
+      {
+         if (!ContainsRange(firstIndex, size))
+            Reset(firstIndex, size);
+
+         // We may read a sub range of the currently available range
+         auto offset = firstIndex.GetIndex() - fFirstIndex.GetIndex();
+
+         if (fNValidValues == fSize)
+            return GetValuePtrAt(offset);
+
+         RBulkSpec bulkSpec;
+         bulkSpec.fFirstIndex = firstIndex;
+         bulkSpec.fCount = size;
+         bulkSpec.fMaskReq = maskReq;
+         bulkSpec.fMaskAvail = &fMaskAvail[offset];
+         bulkSpec.fValues = GetValuePtrAt(offset);
+         auto nRead = fField->ReadBulk(bulkSpec);
+         if (nRead == RBulkSpec::kAllSet) {
+            if ((offset == 0) && (size == fSize)) {
+               fNValidValues = fSize;
+            } else {
+               CountValidValues();
+            }
+         } else {
+            fNValidValues += nRead;
+         }
```

Comment:
if `ReadBulk` returns the number of values for which the mask was flipped from false to true, could these lines just be 
```suggestion
         fNValidValues += nRead;
```
instead?

Reply:
Connected to the question below: At least in principle it is possible that some of the values are already available. If we simply return `bulkSpec.fCount`, we would double count these bits.  On the other hand, I would like to avoid counting the bits in the mask for the common cases. Hence the more complicated logic (which has indeed some smell to it).

Reply:
I think I get what you mean but it does not seem possible in practice that kAll is returned and some of the values were already available: kAll is used by implementations of bulk reading that always read all values.